File: .a-new-structure/b-word/L0-universal/c-hybrid/core/foundation/go/types/workflow.go
//omni:code --go -library
//omni:key B-L0-foundation-types-workflow
//omni:version a-02.00

// ============================================================================
// METADATA
// ============================================================================
//
// Key: B-L0-foundation-types-workflow
// Purpose: Universal workflow operation types for step-based execution
// Biblical: Proverbs 16:9 - "A man's heart deviseth his way"
// Authors: Nova Dawn
// Version: 2.0.0
// Created: 2026-02-16
// Updated: 2026-02-18
//
// Extracted from L3-cpisi types.go — these are generic workflow tracking
// types usable by any layer. CPI-SI specific types (Choice, Halt, Tendency,
// Hebrew states, K:ALIGN) remain in L3.
//
// ============================================================================

package types

// ──────────────────────────────────────────────────────────────────────────
// Imports (in METADATA — Go requires all imports before any declarations)
// ──────────────────────────────────────────────────────────────────────────

import (
	"fmt"

	"creativeworkzstudio.com/bereshit/L0-universal/b-spiral/identity"
)

// Pragma — OmniCode file identity (I1-I4). Runtime-queryable via PragmaGet().
//
// NOTE: This file carries the identity vars for the entire types package.
// Go requires unique package-level names, and sessioncontext.go shares this
// package. The Pragma/Metadata here represent workflow.go specifically;
// sessioncontext.go registers its own identity in init_sessioncontext.go
// or documents its key in its METADATA comment block.
var Pragma = [][2]string{
	// I1: Core — key, format, from, at
	{"I1.key", "B-L0-foundation-types-workflow"},
	{"I1.format", "go"},
	{"I1.from", "b-word/seed/code/L0/go/library.go"},
	{"I1.at", "a-02.00"},
	// I2: Family — type, structure, subtype, role
	{"I2.type", "code"},
	{"I2.structure", "4-block"},
	{"I2.subtype", "library"},
	{"I2.role", "foundation"},
	// I3: Instance — file, title, component, path, provides, brief
	{"I3.file", "workflow.go"},
	{"I3.title", "Workflow Operation Types"},
	{"I3.component", "types"},
	{"I3.path", "L0-universal/c-hybrid/core/foundation/go/types/workflow.go"},
	{"I3.provides", "Universal workflow operation types for step-based execution"},
	{"I3.brief", "Workflow operations, steps, runtime tracking, progress"},
	// I4: Architecture — layer, position, pattern
	{"I4.layer", "L0-universal"},
	{"I4.position", "c-hybrid/core/foundation"},
	{"I4.pattern", "value-types"},
}

// Metadata — OmniCode file metadata (C1-C7). Runtime-queryable via MetadataGet().
var Metadata = [][2]string{
	// C1: State — version, status, created, updated
	{"C1.version", "a-02.00"},
	{"C1.status", "Active"},
	{"C1.created", "2026-02-16"},
	{"C1.updated", "2026-02-18"},
	// C2: Attribution
	{"C2.organization", "CreativeWorkzStudio LLC"},
	{"C2.architect", "Seanje Lenox-Wise"},
	{"C2.implementation", "Nova Dawn"},
	{"C2.copyright", "2025-2026 CreativeWorkzStudio LLC"},
	// C3: Grounding
	{"C3.scripture", "Proverbs 16:9 — A man's heart deviseth his way"},
	{"C3.principle", "Workflow steps trace a path — the LORD directs it"},
	{"C3.anchor", "Genesis 1:1 — In the beginning God created the heaven and the earth"},
	// C4: Dependencies
	{"C4.requires.stdlib", "fmt"},
	{"C4.requires.external", "none"},
	{"C4.requires.internal", "L0-universal/b-spiral/identity"},
	{"C4.consumers", "workflow engine, CLI tools, hooks"},
	{"C4.integration", "import creativeworkzstudio.com/bereshit/L0-universal/c-hybrid/core/foundation/go/types"},
	{"C4.if_missing", "callers must define their own workflow tracking types"},
	// C5: Intent
	{"C5.purpose", "Standardized workflow tracking across all layers"},
	{"C5.philosophy", "Steps are the source of truth — CurrentStep/TotalSteps are convenience"},
	// C6: Roadmap
	{"C6.current", "a-02.00 — Workflow ops, steps, runtime tracking, progress, Display"},
	{"C6.planned", "Workflow persistence, step dependencies, rollback"},
	{"C6.limitations", "No persistence — callers handle I/O"},
	// C7: Classification
	{"C7.tags", "workflow, operation, steps, progress, runtime, types"},
	{"C7.category", "Foundation"},
	{"C7.domain", "types"},
	{"C7.paradigm", "CPI-SI"},
}

// ============================================================================
// END METADATA
// ============================================================================

// ============================================================================
// SETUP
// ============================================================================

// --- Constants ---

// WorkflowOperation represents the type of operation being performed.
// Maps to workflow/*.jsonc files.
type WorkflowOperation string

const (
	WorkflowNone         WorkflowOperation = "none"          // No active workflow
	WorkflowFileCreation WorkflowOperation = "file-creation" // Creating new files from templates
	WorkflowAlignment    WorkflowOperation = "alignment"     // Aligning documentation with reality
	WorkflowBlockUpdate  WorkflowOperation = "block-update"  // Updating existing files block by block
	WorkflowCommit       WorkflowOperation = "commit"        // Git commit workflow
)

// --- Core Types ---

// WorkflowStep represents a step within a workflow.
type WorkflowStep struct {
	Index       int    `json:"index"`       // Step number (1-based)
	Action      string `json:"action"`      // What to do
	Description string `json:"description"` // Human-readable description
	Completed   bool   `json:"completed"`   // Whether step is done
}

// RuntimeWorkflow tracks current workflow state.
//
// Steps is the source of truth for completion — not CurrentStep/TotalSteps.
// CurrentStep indicates which step is actively being worked on.
// TotalSteps is derived from len(Steps) for convenience.
type RuntimeWorkflow struct {
	Operation   WorkflowOperation `json:"operation"`    // Current operation type
	CurrentStep int               `json:"current_step"` // Current step index (1-based)
	TotalSteps  int               `json:"total_steps"`  // Total steps in workflow
	StartedAt   string            `json:"started_at"`   // When workflow started (ISO 8601)
	Steps       []WorkflowStep    `json:"steps"`        // All steps with completion status
}

// ============================================================================
// END SETUP
// ============================================================================

// ============================================================================
// BODY
// ============================================================================

// ──────────────────────────────────────────────────────────────────────────
// 1. Org Chart
// ──────────────────────────────────────────────────────────────────────────
//
// Identity:
//   init, PragmaGet, MetadataGet
//
// Constructors:  NewRuntimeWorkflow
// Queries:       IsActive, IsComplete, Progress, CompletedCount, RemainingCount
// Mutators:      CompleteStep
//
// Flow: NewRuntimeWorkflow → IsActive → (CompleteStep → Progress) → IsComplete

// --- Identity Access ---

// init registers this package in the OmniCode identity registry.
func init() {
	key, ok := PragmaGet("I1.key")
	if !ok {
		return // identity not registerable without key
	}
	identity.Register(key, Pragma, Metadata)
}

// PragmaGet looks up a value from the Pragma identity.
// Returns the value and true if found, empty string and false otherwise.
func PragmaGet(key string) (string, bool) {
	for _, pair := range Pragma {
		if pair[0] == key {
			return pair[1], true
		}
	}
	return "", false
}

// MetadataGet looks up a value from the Metadata.
// Returns the value and true if found, empty string and false otherwise.
func MetadataGet(key string) (string, bool) {
	for _, pair := range Metadata {
		if pair[0] == key {
			return pair[1], true
		}
	}
	return "", false
}

// ──────────────────────────────────────────────────────────────────────────
// 2. Helpers
// ──────────────────────────────────────────────────────────────────────────
// [Reserved: No internal helpers — types package uses only stdlib]

// ──────────────────────────────────────────────────────────────────────────
// 3. Core Operations
// ──────────────────────────────────────────────────────────────────────────

// NewRuntimeWorkflow creates a workflow from an operation and a list of steps.
//
// Steps are indexed 1-based. TotalSteps is derived from len(steps).
// CurrentStep starts at 1 (first step).
func NewRuntimeWorkflow(op WorkflowOperation, startedAt string, steps []WorkflowStep) RuntimeWorkflow {
	indexed := make([]WorkflowStep, len(steps))
	for i, s := range steps {
		indexed[i] = WorkflowStep{
			Index:       i + 1,
			Action:      s.Action,
			Description: s.Description,
			Completed:   s.Completed,
		}
	}
	return RuntimeWorkflow{
		Operation:   op,
		CurrentStep: 1,
		TotalSteps:  len(indexed),
		StartedAt:   startedAt,
		Steps:       indexed,
	}
}

// --- Queries ---

// IsNone returns true if this is the "none" operation.
func (op WorkflowOperation) IsNone() bool {
	return op == WorkflowNone
}

// IsActive returns whether a workflow is currently in progress.
func (w *RuntimeWorkflow) IsActive() bool {
	return !w.Operation.IsNone()
}

// IsComplete returns whether all steps are done.
//
// Steps slice is the source of truth — every step must be marked Completed.
// An empty steps slice with no active operation is considered complete.
func (w *RuntimeWorkflow) IsComplete() bool {
	if len(w.Steps) == 0 {
		return !w.IsActive()
	}
	for _, s := range w.Steps {
		if !s.Completed {
			return false
		}
	}
	return true
}

// Progress returns completion percentage (0-100).
//
// Based on the Steps slice, not CurrentStep/TotalSteps.
func (w *RuntimeWorkflow) Progress() int {
	if len(w.Steps) == 0 {
		return 0
	}
	return w.CompletedCount() * 100 / len(w.Steps)
}

// CompletedCount returns the number of completed steps.
func (w *RuntimeWorkflow) CompletedCount() int {
	count := 0
	for _, s := range w.Steps {
		if s.Completed {
			count++
		}
	}
	return count
}

// RemainingCount returns the number of incomplete steps.
func (w *RuntimeWorkflow) RemainingCount() int {
	return len(w.Steps) - w.CompletedCount()
}

// --- Mutators ---

// CompleteStep marks a step as completed by its 1-based index.
//
// Returns false if the index is out of range.
// Advances CurrentStep to the next incomplete step.
func (w *RuntimeWorkflow) CompleteStep(index int) bool {
	if index < 1 || index > len(w.Steps) {
		return false
	}
	w.Steps[index-1].Completed = true

	// Advance CurrentStep past completed steps
	for w.CurrentStep <= len(w.Steps) && w.Steps[w.CurrentStep-1].Completed {
		w.CurrentStep++
	}
	return true
}

// ──────────────────────────────────────────────────────────────────────────
// 4. Error Handling
// ──────────────────────────────────────────────────────────────────────────
// [Reserved: Types package — CompleteStep returns bool for out-of-range]

// ──────────────────────────────────────────────────────────────────────────
// 5. Public APIs
// ──────────────────────────────────────────────────────────────────────────

// --- String Methods ---

// String returns the workflow operation name.
func (op WorkflowOperation) String() string {
	return string(op)
}

// String returns a human-readable representation of the step.
func (s WorkflowStep) String() string {
	mark := " "
	if s.Completed {
		mark = "x"
	}
	return fmt.Sprintf("[%s] %d. %s", mark, s.Index, s.Action)
}

// String returns a human-readable summary of the workflow.
//
//	"none (inactive)"
//	"file-creation: 2/5 steps (40%)"
//	"alignment: complete (5/5 steps)"
func (w *RuntimeWorkflow) String() string {
	if !w.IsActive() {
		return fmt.Sprintf("%s (inactive)", w.Operation)
	}
	if w.IsComplete() {
		return fmt.Sprintf("%s: complete (%d/%d steps)", w.Operation, w.CompletedCount(), len(w.Steps))
	}
	return fmt.Sprintf("%s: %d/%d steps (%d%%)", w.Operation, w.CompletedCount(), len(w.Steps), w.Progress())
}

// ============================================================================
// END BODY
// ============================================================================

// ============================================================================
// CLOSING
// ============================================================================
//
// ──────────────────────────────────────────────────────────────────────────
// Cv — Validation
// ──────────────────────────────────────────────────────────────────────────
//
// Testing: go test -v -run TestWorkflow ./L0-universal/.../types/
//
// Verify:
//   - WorkflowNone serializes as "none" (not empty string)
//   - IsComplete uses Steps slice, not CurrentStep > TotalSteps
//   - CompleteStep advances CurrentStep correctly
//   - Progress matches CompletedCount / len(Steps) * 100
//
// ──────────────────────────────────────────────────────────────────────────
// Ce — Execution
// ──────────────────────────────────────────────────────────────────────────
//
// Library — imported, not executed.
// import "creativeworkzstudio.com/bereshit/L0-universal/c-hybrid/core/foundation/go/types"
//
// ──────────────────────────────────────────────────────────────────────────
// Cc — Cleanup
// ──────────────────────────────────────────────────────────────────────────
//
// No resources to clean up. Pure value types.
//
// ──────────────────────────────────────────────────────────────────────────
// X1: Policy
// ──────────────────────────────────────────────────────────────────────────
//
// Never:
//   - Change WorkflowNone from "none" to empty string — breaks serde
//   - Use CurrentStep > TotalSteps for completion — use Steps slice
//   - Add filesystem I/O to this package — types only
//
// Careful:
//   - Adding new WorkflowOperation values — update Rust mirror
//   - Changing WorkflowStep fields — breaks JSON contract
//
// Safe:
//   - Adding new methods on existing types
//   - Adding new query functions
//
// ──────────────────────────────────────────────────────────────────────────
// X5: Note
// ──────────────────────────────────────────────────────────────────────────
//
// Note: Generic workflow types extracted from L3-cpisi. Steps slice is the
// source of truth for completion. CurrentStep/TotalSteps are convenience fields.
//
// Scripture: Proverbs 16:9 — "A man's heart deviseth his way: but the LORD
// directeth his steps."
//
// ============================================================================
// END CLOSING
// ============================================================================

File: .a-new-structure/b-word/L0-universal/c-hybrid/core/foundation/src/workflow.rs
//omni:code --rust -library
//omni:key B-L0-foundation-workflow-rs
//omni:version b-02.00

// ============================================================================
// METADATA
// ============================================================================
//
// Key:     B-L0-foundation-workflow-rs
// Purpose: Universal workflow operation types for step-based execution
// Biblical: Proverbs 16:9 — A man's heart deviseth his way: but the LORD
//           directeth his steps
// Version: b-02.00

//! # Workflow Types
//!
//! Universal workflow operation types for step-based execution tracking.
//!
//! | Type | Models | When |
//! |------|--------|------|
//! | [`WorkflowOperation`] | Type of workflow in progress | Enum discriminant |
//! | [`WorkflowStep`] | A single step with action + status | Step tracking |
//! | [`RuntimeWorkflow`] | Full workflow state machine | Orchestration |
//!
//! Extracted from L3-cpisi — these are generic workflow tracking types
//! usable by any layer. CPI-SI specific types remain in L3.

// ============================================================================
// END METADATA
// ============================================================================

// ============================================================================
// SETUP
// ============================================================================

// ──────────────────────────────────────────────────────────────────────────
// 1. Imports
// ──────────────────────────────────────────────────────────────────────────

use serde::{Deserialize, Serialize};
use std::fmt;

// ──────────────────────────────────────────────────────────────────────────
// 7. Core Types
// ──────────────────────────────────────────────────────────────────────────

/// Type of workflow operation being performed.
///
/// Each variant maps to a specific multi-step process. The serialized
/// form uses kebab-case strings for JSON readability.
///
/// # Examples
///
/// ```
/// use bereshit_l0_foundation::WorkflowOperation;
///
/// let op = WorkflowOperation::default();
/// assert_eq!(op, WorkflowOperation::None);
/// assert_eq!(op.to_string(), "none");
/// ```
#[derive(Debug, Clone, Default, PartialEq, Eq, Serialize, Deserialize)]
pub enum WorkflowOperation {
    /// No active workflow.
    #[serde(rename = "none")]
    #[default]
    None,
    /// Creating new files from templates.
    #[serde(rename = "file-creation")]
    FileCreation,
    /// Aligning documentation with reality.
    #[serde(rename = "alignment")]
    Alignment,
    /// Updating existing files block by block.
    #[serde(rename = "block-update")]
    BlockUpdate,
    /// Git commit workflow.
    #[serde(rename = "commit")]
    Commit,
}

/// A step within a workflow.
///
/// Steps are 1-indexed. Each step has an action (machine-readable verb)
/// and a description (human-readable explanation).
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct WorkflowStep {
    /// Step number (1-based).
    pub index: usize,
    /// Machine-readable action verb (e.g., "copy", "lint", "commit").
    pub action: String,
    /// Human-readable description of what this step does.
    pub description: String,
    /// Whether this step has been completed.
    pub completed: bool,
}

/// Tracks current workflow state — the state machine.
///
/// Holds the operation type, current position, and all steps with their
/// completion status. Progress is computed from the steps vec, not from
/// `current_step` — the steps are the source of truth.
///
/// # Examples
///
/// ```
/// use bereshit_l0_foundation::{RuntimeWorkflow, WorkflowOperation, WorkflowStep};
///
/// let mut w = RuntimeWorkflow::new(WorkflowOperation::Alignment, vec![
///     WorkflowStep { index: 1, action: "lint".into(), description: "Run linter".into(), completed: false },
///     WorkflowStep { index: 2, action: "fix".into(), description: "Apply fixes".into(), completed: false },
/// ]);
/// assert!(w.is_active());
/// assert_eq!(w.progress(), 0);
///
/// w.steps[0].completed = true;
/// w.current_step = 2;
/// assert_eq!(w.progress(), 50);
/// ```
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct RuntimeWorkflow {
    /// Current operation type.
    pub operation: WorkflowOperation,
    /// Current step index (1-based). 0 = not started.
    pub current_step: usize,
    /// Total steps in this workflow.
    pub total_steps: usize,
    /// ISO timestamp when workflow started (empty = not started).
    #[serde(default)]
    pub started_at: String,
    /// All steps with completion status.
    #[serde(default)]
    pub steps: Vec<WorkflowStep>,
}

// ============================================================================
// END SETUP
// ============================================================================

// ============================================================================
// BODY
// ============================================================================
//
// Subsection order:
//   3. Constructors   — Creating workflow instances
//   5. Queries        — State inspection
//   6. Output         — Display formatting

// ──────────────────────────────────────────────────────────────────────────
// 3. Constructors
// ──────────────────────────────────────────────────────────────────────────

impl RuntimeWorkflow {
    /// Create a new workflow with the given operation and steps.
    ///
    /// Sets `total_steps` from the steps vec length. `current_step`
    /// starts at 1 (first step) if steps are provided, 0 otherwise.
    pub fn new(operation: WorkflowOperation, steps: Vec<WorkflowStep>) -> Self {
        let total = steps.len();
        Self {
            operation,
            current_step: if total > 0 { 1 } else { 0 },
            total_steps: total,
            started_at: String::new(),
            steps,
        }
    }
}

impl Default for RuntimeWorkflow {
    fn default() -> Self {
        Self {
            operation: WorkflowOperation::None,
            current_step: 0,
            total_steps: 0,
            started_at: String::new(),
            steps: Vec::new(),
        }
    }
}

// ──────────────────────────────────────────────────────────────────────────
// 5. Queries & Accessors
// ──────────────────────────────────────────────────────────────────────────

impl RuntimeWorkflow {
    /// Whether a workflow is currently in progress.
    pub fn is_active(&self) -> bool {
        self.operation != WorkflowOperation::None
    }

    /// Whether all steps are completed.
    ///
    /// A workflow is complete when every step in the steps vec is marked
    /// completed. An empty workflow (no steps) is not complete — it hasn't
    /// started.
    pub fn is_complete(&self) -> bool {
        !self.steps.is_empty() && self.steps.iter().all(|s| s.completed)
    }

    /// Completion percentage (0-100).
    ///
    /// Based on the number of completed steps in the steps vec.
    /// Returns 0 for empty workflows.
    pub fn progress(&self) -> usize {
        if self.total_steps == 0 {
            return 0;
        }
        let completed = self.steps.iter().filter(|s| s.completed).count();
        completed * 100 / self.total_steps
    }

    /// Number of completed steps.
    pub fn completed_count(&self) -> usize {
        self.steps.iter().filter(|s| s.completed).count()
    }

    /// Number of remaining (incomplete) steps.
    pub fn remaining_count(&self) -> usize {
        self.steps.iter().filter(|s| !s.completed).count()
    }
}

impl WorkflowOperation {
    /// Whether this is the `None` variant.
    pub fn is_none(&self) -> bool {
        *self == Self::None
    }
}

// ──────────────────────────────────────────────────────────────────────────
// 6. Output & Display
// ──────────────────────────────────────────────────────────────────────────

impl fmt::Display for WorkflowOperation {
    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
        match self {
            Self::None => write!(f, "none"),
            Self::FileCreation => write!(f, "file-creation"),
            Self::Alignment => write!(f, "alignment"),
            Self::BlockUpdate => write!(f, "block-update"),
            Self::Commit => write!(f, "commit"),
        }
    }
}

impl fmt::Display for RuntimeWorkflow {
    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
        if self.operation == WorkflowOperation::None {
            write!(f, "[idle]")
        } else {
            write!(
                f,
                "[{}] step {}/{} ({}%)",
                self.operation,
                self.current_step,
                self.total_steps,
                self.progress()
            )
        }
    }
}

// ============================================================================
// END BODY
// ============================================================================

// ============================================================================
// CLOSING
// ============================================================================
//
// "A man's heart deviseth his way: but the LORD directeth his steps."
// — Proverbs 16:9

// ──────────────────────────────────────────────────────────────────────────
// Cv — Closing Validation
// ──────────────────────────────────────────────────────────────────────────

#[cfg(test)]
mod tests {
    use super::*;

    // -- WorkflowOperation ------------------------------------------------

    #[test]
    fn operation_default_is_none() {
        assert_eq!(WorkflowOperation::default(), WorkflowOperation::None);
        assert!(WorkflowOperation::None.is_none());
    }

    #[test]
    fn operation_display() {
        assert_eq!(WorkflowOperation::None.to_string(), "none");
        assert_eq!(WorkflowOperation::FileCreation.to_string(), "file-creation");
        assert_eq!(WorkflowOperation::Alignment.to_string(), "alignment");
        assert_eq!(WorkflowOperation::BlockUpdate.to_string(), "block-update");
        assert_eq!(WorkflowOperation::Commit.to_string(), "commit");
    }

    #[test]
    fn operation_serde_roundtrip() {
        let ops = [
            WorkflowOperation::None,
            WorkflowOperation::FileCreation,
            WorkflowOperation::Alignment,
            WorkflowOperation::BlockUpdate,
            WorkflowOperation::Commit,
        ];
        for op in &ops {
            let json = serde_json::to_string(op).unwrap();
            let restored: WorkflowOperation = serde_json::from_str(&json).unwrap();
            assert_eq!(&restored, op);
        }
    }

    #[test]
    fn operation_none_serializes_as_none_string() {
        let json = serde_json::to_string(&WorkflowOperation::None).unwrap();
        assert_eq!(json, "\"none\"");
    }

    // -- RuntimeWorkflow --------------------------------------------------

    #[test]
    fn default_is_inactive() {
        let w = RuntimeWorkflow::default();
        assert!(!w.is_active());
        assert!(!w.is_complete());
        assert_eq!(w.progress(), 0);
        assert_eq!(w.completed_count(), 0);
        assert_eq!(w.remaining_count(), 0);
    }

    #[test]
    fn new_creates_active_workflow() {
        let steps = vec![
            WorkflowStep {
                index: 1,
                action: "a".into(),
                description: "first".into(),
                completed: false,
            },
            WorkflowStep {
                index: 2,
                action: "b".into(),
                description: "second".into(),
                completed: false,
            },
        ];
        let w = RuntimeWorkflow::new(WorkflowOperation::Alignment, steps);
        assert!(w.is_active());
        assert!(!w.is_complete());
        assert_eq!(w.current_step, 1);
        assert_eq!(w.total_steps, 2);
        assert_eq!(w.progress(), 0);
        assert_eq!(w.remaining_count(), 2);
    }

    #[test]
    fn progress_calculation() {
        let w = RuntimeWorkflow {
            operation: WorkflowOperation::FileCreation,
            current_step: 3,
            total_steps: 4,
            started_at: String::new(),
            steps: vec![
                WorkflowStep { index: 1, action: "a".into(), description: "".into(), completed: true },
                WorkflowStep { index: 2, action: "b".into(), description: "".into(), completed: true },
                WorkflowStep { index: 3, action: "c".into(), description: "".into(), completed: false },
                WorkflowStep { index: 4, action: "d".into(), description: "".into(), completed: false },
            ],
        };
        assert!(w.is_active());
        assert!(!w.is_complete());
        assert_eq!(w.progress(), 50);
        assert_eq!(w.completed_count(), 2);
        assert_eq!(w.remaining_count(), 2);
    }

    #[test]
    fn is_complete_when_all_steps_done() {
        let w = RuntimeWorkflow {
            operation: WorkflowOperation::Commit,
            current_step: 2,
            total_steps: 2,
            started_at: String::new(),
            steps: vec![
                WorkflowStep { index: 1, action: "a".into(), description: "".into(), completed: true },
                WorkflowStep { index: 2, action: "b".into(), description: "".into(), completed: true },
            ],
        };
        assert!(w.is_complete());
        assert_eq!(w.progress(), 100);
    }

    #[test]
    fn display_idle() {
        let w = RuntimeWorkflow::default();
        assert_eq!(w.to_string(), "[idle]");
    }

    #[test]
    fn display_active() {
        let w = RuntimeWorkflow {
            operation: WorkflowOperation::Alignment,
            current_step: 2,
            total_steps: 4,
            started_at: String::new(),
            steps: vec![
                WorkflowStep { index: 1, action: "a".into(), description: "".into(), completed: true },
                WorkflowStep { index: 2, action: "b".into(), description: "".into(), completed: false },
                WorkflowStep { index: 3, action: "c".into(), description: "".into(), completed: false },
                WorkflowStep { index: 4, action: "d".into(), description: "".into(), completed: false },
            ],
        };
        assert_eq!(w.to_string(), "[alignment] step 2/4 (25%)");
    }

    #[test]
    fn workflow_serde_roundtrip() {
        let w = RuntimeWorkflow::new(
            WorkflowOperation::BlockUpdate,
            vec![WorkflowStep {
                index: 1,
                action: "lint".into(),
                description: "Run linter".into(),
                completed: false,
            }],
        );
        let json = serde_json::to_string(&w).unwrap();
        let restored: RuntimeWorkflow = serde_json::from_str(&json).unwrap();
        assert_eq!(restored.operation, WorkflowOperation::BlockUpdate);
        assert_eq!(restored.total_steps, 1);
        assert_eq!(restored.steps.len(), 1);
    }
}

// ──────────────────────────────────────────────────────────────────────────
// Ce — Closing Execution
// ──────────────────────────────────────────────────────────────────────────
//
// Library module — no entry point.

// ──────────────────────────────────────────────────────────────────────────
// Cc — Closing Cleanup
// ──────────────────────────────────────────────────────────────────────────
//
// Serde types — no resources to clean up.

//
// ──────────────────────────────────────────────────────────────────────────
// X1: Modification Policy
// ──────────────────────────────────────────────────────────────────────────
//
// Safe to Modify:
//   - Add new WorkflowOperation variants (additive)
//   - Add new query methods on RuntimeWorkflow
//   - Add tests
//
// Modify with Care:
//   - Existing variant serde names — breaks JSON compatibility
//   - WorkflowStep fields — breaks construction
//   - RuntimeWorkflow fields — breaks deserialization
//
// Never Modify:
//   - WorkflowOperation::None as default

// ──────────────────────────────────────────────────────────────────────────
// X5: Closing Note
// ──────────────────────────────────────────────────────────────────────────
//
// Generic workflow types — extracted from L3, usable by all layers.
// "A man's heart deviseth his way: but the LORD directeth his steps."
// — Proverbs 16:9

// ============================================================================
// END CLOSING
// ============================================================================
